FWCore/Framework/src/EventProcessor.cc

```diff
@@ -1458,23 +1478,22 @@ namespace edm {
     FDEBUG(1) << "\tdeleteRunFromCache " << run << "\n";
   }
 
-  void EventProcessor::writeLumiAsync(WaitingTaskHolder task, std::shared_ptr<LuminosityBlockProcessingStatus> iStatus) {
```

Comment:
What is stopping the LuminosityBlock from being deleted before the task created by writeLumiAsync finishes?

Reply:
Couldn't the shared_ptr be destroyed, the LuminosityBlock be returned the the ReusableObjectHolder and another lumi transition pick that object up and reset it while the writLumi is happening?

Reply:
Doesn't the LuminosityBlockProcessingStatus also prevent the LimitedTaskQueue from starting a new task? If the status object is deleted before writeLumi finishes, then another Lumi will start before we've finished with the one we are processing.

Reply:
I see now, the task passed into writeLumiAsync actually is keeping LuminosityBlockProcessingStatus alive. I withdraw my objection.

Reply:
If you look in EventProcessor::globalEndLumiAsync at the first make_waiting_task (which is the one I did not modify), inside that task is a line "status.reset();". Until that point there is a shared_ptr pointer holding onto the status object somewhere and keeping it alive. Part of the time they are in a vector held as a data member by EventProcessor. A lot of other shared_ptr's are really unnecessary for keeping it alive. Some of what I did was remove some of the unnecessary shared_ptr so I did not have to worry about making sure they got deleted soon enough.

Reply:
Note one thing that complicates this is that the destruction order of captured variables is not specified by the C++ standard. Same for function arguments.

Also note, sometimes I used a move on the shared_ptr so that it was already null when it hit the end of a function or lambda. I had to make a few modifications so I could do the moves.